File: books_webflux/src/main/java/ru/zakat/books_webflux/controller/BookController.java
package ru.zakat.books_webflux.controller;

import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import ru.zakat.books_webflux.model.Book;
import ru.zakat.books_webflux.service.BookService;

@RestController
@RequestMapping("/books")
@RequiredArgsConstructor
public class BookController {

    private final BookService bookService;

    @GetMapping("/{id}")
    public Mono<Book> getBookById(@PathVariable Long id) {
        return bookService.getBookById(id)
                .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.NOT_FOUND, "Bank not found")));
    }

    @GetMapping("/error/{id}")
    public Mono<Book> getBookByIdError(@PathVariable Long id) {
        return bookService
                .getBookById(id)
                .switchIfEmpty(Mono.error(new RuntimeException("Book not found")))
                .onErrorResume(e -> Mono.just(new Book(1L, "Unknown", "Unknown", 0.0)));
    }

    @GetMapping
    public Flux<Book> getAllBooks() {
        return bookService.getAllBooks();
    }

    @PostMapping
    public Mono<Book> createBook(@RequestBody Book book) {
        return bookService.createBook(book);
    }

    @DeleteMapping("/{id}")
    public Mono<Void> deleteBook(@PathVariable Long id) {
        return bookService.deleteBook(id);
    }

    @PutMapping("/{id}")
    public Mono<Book> updateBook(@PathVariable Long id, @RequestBody Book book) {
        return bookService.updateBook(id, book);
    }

    @GetMapping("/rated")
    public Flux<Book> findByRating(@RequestParam Double rating) {
        return bookService.getAllBooks()
                .filter(book -> book.getRating() >= rating)
                .onBackpressureBuffer();
    }
}

File: rsocket-client/src/main/kotlin/com/zakat/rsocket_client/controller/BookController.kt
package com.zakat.rsocket_client.controller

import com.zakat.rsocket_client.model.Book
import org.reactivestreams.Publisher
import org.springframework.messaging.rsocket.RSocketRequester
import org.springframework.messaging.rsocket.retrieveFlux
import org.springframework.messaging.rsocket.retrieveMono
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

@RestController
@RequestMapping("/api/books")
class BookController(
    val rSocketRequester: RSocketRequester
) {

    @GetMapping
    fun getBooks(): Publisher<Book> {
        return rSocketRequester
            .route("getBooks")
            .retrieveFlux()
    }

    @GetMapping("/{bookId}")
    fun getBook(@PathVariable bookId: String): Mono<Book> {
        return rSocketRequester
            .route("getBook")
            .data(bookId)
            .retrieveMono()
    }

    @PostMapping
    fun addBook(@RequestBody book: Book): Mono<Book> {
        return rSocketRequester
            .route("addBook")
            .data(book)
            .retrieveMono()
    }

    @DeleteMapping("/{bookId}")
    fun deleteBook(@PathVariable bookId: String): Publisher<Void> {
        return rSocketRequester
            .route("deleteBook")
            .data(bookId)
            .send()
    }

    @PostMapping("/addBooks")
    fun addBooks(@RequestBody books: List<Book>): Flux<Book> {
        return rSocketRequester
            .route("bookChannel")
            .data(books)
            .retrieveFlux()
    }
}